presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java

```diff
@@ -345,20 +346,20 @@ public Void visitJoin(JoinNode node, Set<Symbol> boundSymbols)
                         predicateSymbols,
                         allInputs);
             });
-
-            checkDependencies(allInputs, node.getOutputSymbols(), "Symbol from join output (%s) not in sources (%s)", node.getOutputSymbols(), allInputs);
-
-            List<Integer> rightSymbolIndexes = node.getRight().getOutputSymbols().stream()
-                    .filter(symbol -> node.getOutputSymbols().contains(symbol))
+            Optional<Integer> lastLeftSymbolIndex = node.getLeft().getOutputSymbols().stream()
                     .map(symbol -> node.getOutputSymbols().indexOf(symbol))
-                    .collect(toImmutableList());
+                    .filter(index -> index >= 0)
+                    .reduce(BinaryOperator.maxBy(Integer::compareTo));
 
-            boolean allLeftBeforeRight = node.getLeft().getOutputSymbols().stream()
-                    .filter(symbol -> node.getOutputSymbols().contains(symbol))
+            Optional<Integer> firstRightSymbolIndex = node.getRight().getOutputSymbols().stream()
                     .map(symbol -> node.getOutputSymbols().indexOf(symbol))
```

Comment:
is `indexOf` O(n)? If so then iterating all n symbols and retrieving their indexes is O(n*n).

Maybe you could do something like:
```
for (int i = 0; i < node.getOutputSymbols().size(); i++) {
  Symbol symbol =  node.getOutputSymbols().get(i);
  if (leftSymbols.contains(symbol)) {
     // update max left Symbol
  } 
  ...
```
?

Reply:
`leftSymbols.contains(symbol)` also performs `indexOf(symbol) > 0` internally wouldn't that be a issue ?  In the initial version of the code

```
List<Integer> rightSymbolIndexes = node.getRight().getOutputSymbols().stream()
     .filter(symbol -> node.getOutputSymbols().contains(symbol))
     .map(symbol -> node.getOutputSymbols().indexOf(symbol))
     .collect(toImmutableList());
````
might ideally do the same job twice

Reply:
I am sorry that I wasn't precise enough. 

`leftSymbols.contains(symbol)` -  I meant using `HashSet` for `leftSymbols` here.

Reply:
Oh yes that can be done